[mlir][Target] Improve ROCDL gpu serialization API

fabianmcg opened the pull request:
This patch improves the ROCDL gpu serialization API by:
- Introducing the enum `AMDGCNLibraries` for specifying the AMD GCN device code libraries to use during linking.
- Removing `getCommonBitcodeLibs` in favor of `AMDGCNLibraries`. Previously `getCommonBitcodeLibs` would try to load all AMD GCN bitcode librariesm now it will only load the requested libraries.
- Exposing the `compileToBinary` method and making it virtual, allowing downstream users to re-use this method.
- Exposing `moduleToObjectImpl`, this method provides a prototype flow for compiling to binary, allowing downstream users to re-use this method.
- It also avoids constructing the control variables if no device libraries are being used.

This patch also changes the behavior of the CMake flag `DEFAULT_ROCM_PATH`. Before it would fall back to a default value of `/opt/rocm` if not specified. However, that default value causes fragile builds in environments with ROCm. Now, the flag falls back to the empty string, making it clear that **the user must provide a value at LLVM build time**.

fabianmcg committed:
[mlir][Target] ROCDL

krzysz00 commented on mlir/include/mlir/Target/LLVM/ROCDL/Utils.h:
```diff
@@ -27,6 +27,62 @@ namespace ROCDL {
 
+/// Helper class for specifying the AMD GCN device libraries required for
+/// compilation.
+class AMDGCNLibraryList {
```
Nit, isn't this just `enum class AMDGCNLibraryList : uint32_t`?

... Actually, on top of that - there's already general support for bit enums. This could and should be tablegen'd

fabianmcg replied:
I was trying to avoid going to tablegen for such a small class, but I'll switch it.

krzysz00 replied:
It's more that there's a lot of common infrastructure around these sorts of flag enums (including, say, printing them) that can be autogenerated. And it'll be useful to have if someone ever wants to put this as an attribute somewhere

krzysz00 commented on mlir/include/mlir/Target/LLVM/ROCDL/Utils.h:
```diff
@@ -27,6 +27,62 @@ namespace ROCDL {
+  }
+
+  /// Adds all the libraries in `list` to the library list.
+  AMDGCNLibraryList addList(AMDGCNLibraryList list) {
```
For instance, this is `bitEnumSet()` or some similarly-phrased function

krzysz00 commented on mlir/lib/Target/LLVM/CMakeLists.txt:
```diff
@@ -123,17 +123,12 @@ add_mlir_dialect_library(MLIRROCDLTarget
   elseif(DEFINED ENV{ROCM_PATH})
     set(DEFAULT_ROCM_PATH "$ENV{ROCM_PATH}" CACHE PATH "Fallback path to search for ROCm installs")
   else()
-    set(DEFAULT_ROCM_PATH "/opt/rocm" CACHE PATH "Fallback path to search for ROCm installs")
```
This is meant to be a fallback so that the build goes to look in `/opt/rocm` for the device libraries if there are no clues about where they are in the user's environment, though?

fabianmcg replied:
Yes, and the flag is still available, however, having a hard coded value causes issues in windows, as the value is for linux. a better solution would be detecting it with CMake, I'll look into if there's anything like findHip in CMake.

krzysz00 commented on mlir/lib/Target/LLVM/CMakeLists.txt:
```diff
@@ -123,17 +123,12 @@ add_mlir_dialect_library(MLIRROCDLTarget
 
 if(MLIR_ENABLE_ROCM_CONVERSIONS)
-  if (NOT ("AMDGPU" IN_LIST LLVM_TARGETS_TO_BUILD))
-    message(SEND_ERROR
```
Is this because there's a duplicate check? What happens if the AMDGPU backend isn't available?

fabianmcg replied:
MLIR_ENABLE_ROCM_CONVERSIONS is already an alias for checking whether AMDGPU is being built or not.

krzysz00 reviewed (commented):
I've got a few notes

fabianmcg committed:
switch enum over to enum class

fabianmcg commented:
@krzysz00 I'm still not convinced on using tablegen for the AMD GCN lib enum as there are no consumers for it as an attribute, and we generally keep things minimal and internal unless there's a use case, and right now there are none. Also, I would argue that the enum in question is completely internal to the compilation process and it shouldn't be exposed.
However, I switched the class to an `enum class`, and by using https://www.llvm.org/doxygen/BitmaskEnum_8h.html I removed many of the methods, reducing code duplication.

With respect to CMake and `DEFAULT_ROCM_PATH`, right now using `FindHip` is not robust enough, as the module still lives inside `/opt/rocm` and not CMake. Hence, I'm arguing that it's the responsibility of the builder/maintainer to specify `DEFAULT_ROCM_PATH` when building LLVM, otherwise we will encounter fragile builds and issues with windows.

krzysz00 reviewed (approved):
These all seem like reasonable changes to me, approved
